src/convert/replacements.ts

```diff
@@ -0,0 +1,203 @@
+  const filenames = [cmp.xml, ...cmp.walkContent()].filter(Boolean);
+  const replacementsForComponent = (
+    await Promise.all(
+      // build a nested array that can be run through Object.fromEntries
+      // one MarkedReplacement[] for each file in the component
+      filenames.map(
+        async (f): Promise<[string, MarkedReplacement[]]> => [
+          f,
+          await Promise.all(
+            replacementConfigs
+              // filter out any that don't match the current file
+              .filter((r) => matchesFile(f, r))
+              .map(async (r) => ({
+                // used during replacement stream to limit warnings to explicit filenames, not globs
+                singleFile: Boolean(r.filename),
+                // Config is json which might use the regex.  If so, turn it into an actual regex
+                toReplace: r.stringToReplace ?? new RegExp(r.regexToReplace, 'g'),
+                // get the literal replacement (either from env or file contents)
+                replaceWith: r.replaceWithEnv
+                  ? getEnvValue(r.replaceWithEnv)
+                  : await getContentsOfReplacementFile(r.replaceWithFile),
+              }))
+          ),
+        ]
+      )
+    )
+  )
+    // filter out any that don't have any replacements
+    .filter(([, replacements]) => replacements.length > 0);
+
+  if (replacementsForComponent.length) {
+    // turn into a Dictionary-style object so it's easier to lookup by filename
+    return Object.fromEntries(replacementsForComponent);
+  }
+};
+
+export const matchesFile = (f: string, r: ReplacementConfig): boolean =>
+  // filenames will be absolute.  We don't have convenient access to the pkgDirs,
+  // so we need to be more open than an exact match
+  f.endsWith(r.filename) || (r.glob && minimatch(f, `**/${r.glob}`));
+
+/**
+ * Regardless of any components, return the ReplacementConfig that are valid with the current env.
+ * These can be checked globally and don't need to be checked per component.
+ */
+const envFilter = (replacementConfigs: ReplacementConfig[] = []): ReplacementConfig[] =>
+  replacementConfigs.filter(
+    (replacement) =>
+      !replacement.replaceWhenEnv ||
+      replacement.replaceWhenEnv.every((envConditional) => process.env[envConditional.env] === envConditional.value)
+  );
+
+/** A "getter" for envs to throw an error when an expected env is not present */
+const getEnvValue = (env: string): string => {
+  if (process.env[env]) {
+    return process.env[env];
+  }
+  throw new SfError(
+    `"${env}" is in sfdx-project.json as a value for "replaceWithEnv" property, but it's not set in your environment.`
+  );
+};
+
+/**
+ * Read the `replacement` property from sfdx-project.json
+ */
+const readReplacementsFromProject = async (): Promise<ReplacementConfig[]> => {
+  const proj = await SfProject.resolve();
+  const projJson = (await proj.resolveProjectConfig()) as { replacements?: ReplacementConfig[] };
```

Comment:
@mshanemc is your work going to include updating sfdx-project.schema.json in @salesforce/schemas?

Reply:
it should.  good call.